Replace separate https and force_ssl flags with https + hsts

oliverguenther opened the pull request:
By default, https will be enabled in production in OpenProject 12.2. Right now, a separate flag exists called `rails_force_ssl` that needs to be controlled separately, but the two settings are intermingled.

This PR fixes this so that:

HTTPS setting controls the secure cookies and generating links with https throughout the application

HSTS setting is only active by default when HTTPS is enabled and controls the HSTS headers as well as the HTTP->HTTPS request upgrade for transport-security.

oliverguenther committed:
Replace separate https and force_ssl flags with https + hsts

By default, https will be enabled in production in OpenProject 12.2. Right now, a separate flag exists called

rails_force_ssl that needs to be controlled separately, but the two settings are intermingled.

This PR fixes this so that:

HTTPS setting controls the secure cookies and generating links with
https throughout the application

HSTS setting is only active by default when HTTPS is enabled and
controls the HSTS headers as well as the HTTP->HTTPS request upgrade for
transport-security.

ghost commented:
@oliverguenther, i am just reading and thinking about connection, maybe the packager needs to be adjusted here? is Cyril aware in case he needs to change things? What do you think?

oliverguenther commented:
@adam-op Yes packager and docs still need to be adapted. This will happen in this PR as well. I first need to discuss this change with Jens and Markus

oliverguenther committed:
Change protocol mismatch warning bar to print correct warning

We can only link to the installation guides now that the protocol
setting can no longer be controlled in the UI.

oliverguenther committed:
Rename FORCE_SSL to HSTS in packager setup

oliverguenther committed:
Add documentation on HTTPS/HSTS mode on docker

oliverguenther committed:
Improve documentation on https mode and protocol for packager

oliverguenther committed:
Add separate note that SSL does not mean container is running ssl

oliverguenther committed:
Allow settings defaults to be lambdas even if format present

cbliard commented on docs/installation-and-operations/configuration/environment/README.md:
```diff
@@ -87,7 +87,8 @@ OPENPROJECT_SESSION__COOKIE__NAME (default="_open_project_session")
 OPENPROJECT_RAILS__RELATIVE__URL__ROOT (default="")
-OPENPROJECT_RAILS__FORCE__SSL (default=false)
+OPENPROJECT_HTTPS (default=true when production, false in development)
+OPENPROJECT_HSTS (default=false)
```
```suggestion
OPENPROJECT_HSTS (default=true)
```

cbliard commented on docs/installation-and-operations/installation/docker/README.md:
```diff
@@ -173,6 +181,23 @@ docker run -d -e KEY1=VALUE1 -e KEY2=VALUE2 ...
+
+#### Disabling HSTS headers and insecure request upgrade
+
+If OpenProject is running as HTTPS, it will output HTTP Strict Transport Security (HSTS) headers by default, which will force the browser to upgrade a request to HTTPS even before calling the web server. On the server side, if a request is still unsecured, OpenProject will return a redirect to HTTPS.
```
```suggestion
If OpenProject is running as HTTPS, it will output HTTP Strict Transport Security (HSTS) headers by default, which will force the browser to upgrade a request to HTTPS even before calling the web server. On the server side, if a request is still issued through HTTP, OpenProject will return a redirect to HTTPS.
```

Precise what `unsecured` means

cbliard commented on docs/installation-and-operations/installation/docker/README.md:
````diff
@@ -249,12 +274,9 @@ If you want to run OpenProject in a subdirectory on your server, first you will
 ```
 
-The `force ssl` option can be left out if you are not using HTTPS.
-
````
Nice catch

cbliard commented on docs/installation-and-operations/installation/packaged/README.md:
```diff
@@ -521,6 +527,22 @@ OpenProject heavily relies on caching, which is why the wizard suggests you to i
 
 ![06-cache](06-cache.png)
 
+## Step 7: Host name and Protocol (only when Apache installation was skipped)
```
Thank you for adding documentation for this. It got me confused when I was looking up the code to find where the SSL setting is actually set (because it's not here but in the pkgr apache addon).

cbliard commented on lib/open_project/static/links.rb:
```diff
@@ -98,6 +98,22 @@ def static_links
+            },
+            manual_installation: {
+              href: 'https://www.openproject.org/docs/installation-and-operations/installation/manual/',
+              label: 'Docker installation'
```
```suggestion
              label: 'Manual installation'
```

Well... only `href` is used for this one, but the label did not match

oliverguenther replied:
Good catch, I wanted to add the label just in case it is used somewhere, and it should be correct.

cbliard commented on docs/installation-and-operations/installation/docker/README.md:
```diff
@@ -173,6 +181,23 @@ docker run -d -e KEY1=VALUE1 -e KEY2=VALUE2 ...
+#### Disabling HTTPS mode
+
+By default, OpenProject will expect a HTTPS request in production systems.
+In most cases, you will have an external web server or load balancer terminating the SSL/TLS connection and proxy/reverse-proxy to the docker container. You will then have to set up the web server to forward the protocol information (usually, this is X-Forwarded-Proto but depends on your web server).
```
```suggestion
In most cases, you will have an external web server or load balancer terminating the SSL/TLS connection and proxy/reverse-proxy to the docker container. You will then have to set up the web server to forward the protocol information (usually, this is `X-Forwarded-Proto` but depends on your web server).
```

cbliard commented on docs/installation-and-operations/installation/docker/README.md:
````diff
@@ -78,7 +78,11 @@ The fastest way to get an OpenProject instance up and running is to run the
-docker run -it -p 8080:80 -e OPENPROJECT_SECRET_KEY_BASE=secret -e OPENPROJECT_HOST__NAME=localhost:8080 openproject/community:12
+docker run -it -p 8080:80 \
+  -e OPENPROJECT_SECRET_KEY_BASE=secret \
+  -e OPENPROJECT_HOST__NAME=localhost:8080 \ # Set the host name to what you will externally use it with
+  -e OPENPROJECT_HTTPS=false \ # Disable HTTPS due to accessing container directly
+  openproject/community:12
 ```
 
````
Having `\` to split lines wont work if followed by `#`.

I can't propose a good suggestion because of the triple backticks in the edited code, but I would suggest adding another paragraph after the command explaining why `OPENPROJECT_HOST__NAME=localhost:8080` and `OPENPROJECT_HTTPS=false` are set.

cbliard reviewed (approved):
Code wise, it should be working. I have some minor remarks and improvement propositions here and there.

oliverguenther commented:
Thanks @cbliard for taking the time to go through it. I've in the mean time set up docker with and without https on this branch and this appears to be working well.

cbliard reviewed (approved):
Looks good.

Also I noticed that the documentation about [System Setting > General](https://www.openproject.org/docs/system-admin-guide/system-settings/general-settings/) is not up-to-date as it still documents the Protocol drop-down with HTTPS/HTTP choices. It is also related to [work package #43714](https://community.openproject.org/projects/openproject/work_packages/43714/activity).

Can we merge it anyway so I can test packaged installation as soon as the packages are built?

cbliard commented:
Works with packager too.

torwag commented:
Does this effect also the repro in openproject-deploy? 
I phase exactly the challenge to access a fresh installation over http, as it enforces https

oliverguenther commented:
> Does this effect also the repro in openproject-deploy? I phase exactly the challenge to access a fresh installation over http, as it enforces https

Yes this will be the same issue, I'll make sure to document it there as well. A fixed release will be made tomorrow. Sorry for the wait

oliverguenther commented:
related ticket: https://community.openproject.org/wp/43767

ghost commented:
@oliverguenther could you please explain a readable documentation for containerd docker-compose, i could edit it but I cannot find out, see here for details: https://community.openproject.org/topics/15251?r=15352#message-15352
